File: video/gtk4/examples/gtksink.rs
use gst::prelude::*;

use gtk::prelude::*;
use gtk::{gdk, gio, glib};

use std::cell::RefCell;

fn create_ui(app: &gtk::Application) {
    let pipeline = gst::Pipeline::new();

    let overlay = gst::ElementFactory::make("clockoverlay")
        .property("font-desc", "Monospace 42")
        .build()
        .unwrap();

    let gtksink = gst::ElementFactory::make("gtk4paintablesink")
        .build()
        .unwrap();

    let paintable = gtksink.property::<gdk::Paintable>("paintable");

    // TODO: future plans to provide a bin-like element that works with less setup
    let (src, sink) = if paintable
        .property::<Option<gdk::GLContext>>("gl-context")
        .is_some()
    {
        let src = gst::ElementFactory::make("gltestsrc").build().unwrap();

        let sink = gst::ElementFactory::make("glsinkbin")
            .property("sink", &gtksink)
            .build()
            .unwrap();
        (src, sink)
    } else {
        let src = gst::ElementFactory::make("videotestsrc").build().unwrap();

        let sink = gst::Bin::default();
        let convert = gst::ElementFactory::make("videoconvert").build().unwrap();

        sink.add(&convert).unwrap();
        sink.add(&gtksink).unwrap();
        convert.link(&gtksink).unwrap();

        sink.add_pad(&gst::GhostPad::with_target(&convert.static_pad("sink").unwrap()).unwrap())
            .unwrap();

        (src, sink.upcast())
    };

    pipeline.add_many([&src, &overlay, &sink]).unwrap();
    let caps = gst_video::VideoCapsBuilder::new()
        .width(640)
        .height(480)
        .any_features()
        .build();

    src.link_filtered(&overlay, &caps).unwrap();
    overlay.link(&sink).unwrap();

    let window = gtk::ApplicationWindow::new(app);
    window.set_default_size(640, 480);

    let vbox = gtk::Box::new(gtk::Orientation::Vertical, 0);

    let gst_widget = gstgtk4::RenderWidget::new(&gtksink);
    vbox.append(&gst_widget);

    let label = gtk::Label::new(Some("Position: 00:00:00"));
    vbox.append(&label);

    window.set_child(Some(&vbox));
    window.present();

    app.add_window(&window);

    let pipeline_weak = pipeline.downgrade();
    let timeout_id = glib::timeout_add_local(std::time::Duration::from_millis(500), move || {
        let Some(pipeline) = pipeline_weak.upgrade() else {
            return glib::ControlFlow::Break;
        };

        let position = pipeline.query_position::<gst::ClockTime>();
        label.set_text(&format!("Position: {:.0}", position.display()));
        glib::ControlFlow::Continue
    });

    let bus = pipeline.bus().unwrap();

    pipeline
        .set_state(gst::State::Playing)
        .expect("Unable to set the pipeline to the `Playing` state");

    let app_weak = app.downgrade();
    let bus_watch = bus
        .add_watch_local(move |_, msg| {
            use gst::MessageView;

            let Some(app) = app_weak.upgrade() else {
                return glib::ControlFlow::Break;
            };

            match msg.view() {
                MessageView::Eos(..) => app.quit(),
                MessageView::Error(err) => {
                    println!(
                        "Error from {:?}: {} ({:?})",
                        err.src().map(|s| s.path_string()),
                        err.error(),
                        err.debug()
                    );
                    app.quit();
                }
                _ => (),
            };

            glib::ControlFlow::Continue
        })
        .expect("Failed to add bus watch");

    let timeout_id = RefCell::new(Some(timeout_id));
    let pipeline = RefCell::new(Some(pipeline));
    let bus_watch = RefCell::new(Some(bus_watch));
    app.connect_shutdown(move |_| {
        window.close();

        drop(bus_watch.borrow_mut().take());
        if let Some(pipeline) = pipeline.borrow_mut().take() {
            pipeline
                .set_state(gst::State::Null)
                .expect("Unable to set the pipeline to the `Null` state");
        }

        if let Some(timeout_id) = timeout_id.borrow_mut().take() {
            timeout_id.remove();
        }
    });
}

fn main() -> glib::ExitCode {
    gst::init().unwrap();
    gtk::init().unwrap();

    gstgtk4::plugin_register_static().expect("Failed to register gstgtk4 plugin");

    let app = gtk::Application::new(None::<&str>, gio::ApplicationFlags::default());

    app.connect_activate(create_ui);
    let res = app.run();

    unsafe {
        gst::deinit();
    }

    res
}

File: video/gtk4/examples/gtksink.py
import gi
import sys


gi.require_version('GLib', '2.0')
from gi.repository import GLib
gi.require_version('Gtk', '4.0')
from gi.repository import Gtk
gi.require_version('Gst', '1.0')
from gi.repository import Gst

def on_activate(app):
    pipeline = Gst.Pipeline.new()

    overlay = Gst.ElementFactory.make('clockoverlay', None)
    overlay.set_property('font-desc', 'Monospace 42')

    gtksink = Gst.ElementFactory.make('gtk4paintablesink', None)
    paintable = gtksink.get_property('paintable')

    if paintable.props.gl_context:
        print('Using GL')
        src = Gst.ElementFactory.make('gltestsrc', None)

        sink = Gst.ElementFactory.make('glsinkbin', None)
        sink.set_property('sink', gtksink)
    else:
        print('Not using GL')
        src = Gst.ElementFactory.make('videotestsrc', None)

        sink = Gst.Bin.new()
        convert = Gst.ElementFactory.make('videoconvert', None)

        sink.add(convert)
        sink.add(gtksink)
        convert.link(gtksink)

        sink.add_pad(Gst.GhostPad.new('sink', convert.get_static_pad('sink')))

    pipeline.add(src)
    pipeline.add(overlay)
    pipeline.add(sink)

    src.link_filtered(overlay, Gst.Caps.from_string('video/x-raw(ANY), width=640, height=480'))
    overlay.link(sink)

    window = Gtk.ApplicationWindow(application=app)
    window.set_default_size(640, 480)

    vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)

    picture = Gtk.Picture.new()
    picture.set_paintable(paintable)

    vbox.append(picture)

    label = Gtk.Label.new('Position: 00:00:00')
    vbox.append(label)

    window.set_child(vbox)
    window.present()

    app.add_window(window)

    def on_timeout():
        (res, position) = pipeline.query_position(Gst.Format.TIME)
        if not res or position == -1:
            label.set_text('Position: 00:00:00')
        else:
            seconds = int(position / 1000000000)
            minutes = int(seconds / 60)
            hours = int(minutes / 60)
            seconds = seconds % 60
            minutes = minutes % 60
            label.set_text(f'Position: {hours:02}:{minutes:02}:{seconds:02}')

        return True

    GLib.timeout_add(500, on_timeout)

    pipeline.set_state(Gst.State.PLAYING)

    def on_bus_msg(_, msg):
        match msg.type:
            case Gst.MessageType.EOS:
                app.quit()
            case Gst.MessageType.ERROR:
                (err, _) = msg.parse_error()
                print(f'Error from {msg.src.path_string()}: {err}')
                app.quit()

        return True

    bus = pipeline.get_bus()
    bus.add_watch(GLib.PRIORITY_DEFAULT, on_bus_msg)

    def on_close(_):
        window.close()
        pipeline.set_state(Gst.State.NULL)
    app.connect('shutdown', on_close)

if __name__ == '__main__':
    Gst.init(None)
    app = Gtk.Application()
    app.connect('activate', on_activate)
    res = app.run(None)
    sys.exit(res)
